### ISSUE TITLE
Getting weird errors on autocommands for *.norg

### ISSUE THREAD

**Issue by vhyrro** [nvim-neorg/neorg#74]
Are you experiencing errors after updating the plugin to the latest version, and are getting an error along the lines of:

```lua
E5108: Error executing lua [string "luaeval()"]:1: Vim(echoerr):E121: Undefined variable: Failed                                                                                                                    
Error detected while processing BufEnter Autocommands for "*.norg":                                                                                                                                                 
E5108: Error executing lua Failed to load parser: uv_dlopen: /home/vhyrro/.local/share/nvim/site/pack/packer/start/nvim-treesitter/parser/norg.so: undefined symbol: tree_sitter_norg_external_scanner_create          
Error detected while processing BufEnter Autocommands for "*":                                                                                                                                                      
E5108: Error executing lua Failed to load parser: uv_dlopen: /home/vhyrro/.local/share/nvim/site/pack/packer/start/nvim-treesitter/parser/norg.so: undefined symbol: tree_sitter_norg_external_scanner_create          
Error detected while processing CursorMoved Autocommands for "*":                                                                                                                                                   
E5108: Error executing lua ...start/nvim-treesitter-context/lua/treesitter-context.lua:236: Vim(lua):E5108: Error executing lua Failed to load parser: uv_dlopen: /home/vhyrrolocal/share/nvim/site/pack/packer/start
/nvim-treesitter/parser/norg.so: undefined symbol: tree_sitter_norg_external_scanner_create                                                                                                                         
Error detected while processing WinScrolled Autocommands for "*":                                                                                                                                                   
E5108: Error executing lua Failed to load parser: uv_dlopen: /home/vhyrro/.local/share/nvim/site/pack/packer/start/nvim-treesitter/parser/norg.so: undefined symbol: tree_sitter_norg_external_scanner_create          
Failed to get context: Failed to load parser: uv_dlopen: /home/vhyrro/.local/share/nvim/site/pack/packer/start/nvim-treesitter/parser/norg.so: undefined symbol: tree_sitter_norg_external_scanner_create     
```

Fear no more! Simply take a look at the readme present in the `unstable` branch and scroll down to the treesitter installation section. We've slightly changed the code snippet as the new parser we're using also includes a scanner. After you've taken that code and replaced it in your own config run `:TSUpdate` and voila!

**Comment by vhyrro** [comment]
To all people who still had issues: I think all of those should be fixed now. I downgraded the codebase to C++11 and removed the "modern" [[nodiscard]] attribute with parameters. Let me know if it fixes stuff for y'all and sorry for all the issues, things like that happen with a full parser rewrite :sweat_smile:

**Comment by ryleelyman** [comment]
I reverted the symlink @danymat suggested and tried doing `:TSUpdate norg`. Unfortunately it gives me more or less the same slew of errors as before.

**Comment by wfchandler** [comment]
I'm seeing the same, also on MacOS 11.5 with Clang 12.0.5 aliased as `cc`.  This is a very recent version of Clang and defaults to `-std=gnu17`, so it should support `[[nodiscard]]` by default.

Testing https://github.com/vhyrro/tree-sitter-norg outside of nvim, I can run `yarn gen` successfully, but `yarn test` fails with the same errors as `TSInstall norg`:

<details>
<summary>Expand to view 'yarn test' errors</summary>
<pre>

yarn test
yarn run v1.22.11
$ tree-sitter test
Parser compilation failed.
Stdout: 
Stderr: /Users/wchandler/test/tree-sitter-norg/src/scanner.cc:96:39: error: expected '(' for function-style cast or type construction
        return std::vector<TokenType> { lhs, static_cast<TokenType>(rhs) };
               ~~~~~~~~~~~~~~~~~~~~~~ ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:99:24: warning: rvalue references are a C++11 extension [-Wc++11-extensions]
        std::vector<TokenType>&& operator|(std::vector<TokenType>&& lhs, TokenType rhs)
                              ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:99:59: warning: rvalue references are a C++11 extension [-Wc++11-extensions]
        std::vector<TokenType>&& operator|(std::vector<TokenType>&& lhs, TokenType rhs)
                                                                 ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:322:38: error: expected ';' at end of declaration list
        std::vector<size_t>& get_tag_stack() noexcept { return m_TagStack; }
                                            ^
                                            ;
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:551:23: warning: in-class initialization of non-static data member is a C++11 extension [-Wc++11-extensions]
    bool m_Whitespace = false;
                      ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:555:27: warning: in-class initialization of non-static data member is a C++11 extension [-Wc++11-extensions]
    TokenType m_LastToken = NONE;
                          ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:558:26: warning: in-class initialization of non-static data member is a C++11 extension [-Wc++11-extensions]
    size_t m_ParsedChars = 0, m_IndentationLevel = 0;
                         ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:558:50: warning: in-class initialization of non-static data member is a C++11 extension [-Wc++11-extensions]
    size_t m_ParsedChars = 0, m_IndentationLevel = 0;
                                                 ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:564:60: warning: in-class initialization of non-static data member is a C++11 extension [-Wc++11-extensions]
    const std::array<unsigned char, 6> s_DetachedModifiers = { '*', '-', '>', '|', '=', '~' };
                                                           ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:125:13: error: use of undeclared identifier 'advance'
            advance(lexer);
            ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:134:13: error: use of undeclared identifier 'advance'
            advance(lexer);
            ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:143:5: error: use of undeclared identifier 'advance'
                                advance(lexer);
                                ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:148:6: error: use of undeclared identifier 'advance'
                                        advance(lexer);
                                        ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:159:4: error: use of undeclared identifier 'advance'
                        advance(lexer);
                        ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:163:36: error: use of undeclared identifier 'm_Current'
                                if (lexer->lookahead == ']' && m_Current != '\\')
                                                               ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:165:6: error: use of undeclared identifier 'advance'
                                        advance(lexer);
                                        ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:173:5: error: use of undeclared identifier 'advance'
                                advance(lexer);
                                ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:180:17: error: use of undeclared identifier 'check_link'
                return check_link(lexer);
                       ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:183:10: error: use of undeclared identifier 'advance'
                advance(lexer);
                ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:201:14: error: use of undeclared identifier 'skip'
                skip(lexer);
                ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:206:5: error: use of undeclared identifier 'advance'
                                advance(lexer);
                                ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:212:6: error: use of undeclared identifier 'advance'
                                        advance(lexer);
                                        ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:215:7: error: use of undeclared identifier 'advance'
                                                advance(lexer);
                                                ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:218:8: error: use of undeclared identifier 'advance'
                                                        advance(lexer);
                                                        ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:222:10: error: use of undeclared identifier 'advance'
                                                                        advance(lexer);
                                                                        ^
/Users/wchandler/test/tree-sitter-norg/src/scanner.cc:247:7: error: use of undeclared identifier 'advance'
                                                advance(lexer);
                                                ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
7 warnings and 20 errors generated.

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

</pre>
</details>

The first error looks like the compiler is failing to recognize a braced initializer list, not clear to me why.

**Comment by vhyrro** [comment]
Wow, this is really weird. I might get in touch with the nvim-treesitter guys and ask them if they know of anything that could be done to fix this stuff.

**Comment by Gelio** [comment]
After including the scanner in the list of files, I am getting the following errors in `:messages`:

```
Error detected while processing BufEnter Autocommands for "*":
E5108: Error executing lua ...ejQS/usr/share/nvim/runtime/lua/vim/treesitter/query.lua:161: query: invalid node type at position 50
Error detected while processing CursorMoved Autocommands for "*":
E5108: Error executing lua ...start/nvim-treesitter-context/lua/treesitter-context.lua:236: Vim(lua):E5108: Error executing lua ...ejQS/usr/share/nvim/runtime/lua/vim/treesitter/query.lua:161: query: invalid node type at position 50
Error detected while processing WinScrolled Autocommands for "*":
E5108: Error executing lua ...ejQS/usr/share/nvim/runtime/lua/vim/treesitter/query.lua:161: query: invalid node type at position 50
Failed to get context: ...ejQS/usr/share/nvim/runtime/lua/vim/treesitter/query.lua:161: query: invalid node type at position 50
```

`:checkhealth`:

```
## Installation
  - WARNING: `tree-sitter` executable not found (parser generator, only needed for :TSInstallFromGrammar, not required for :TSInstall)
  - OK: `node` found v14.17.0 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl" }
  - OK: Neovim was compiled with tree-sitter runtime ABI version 13 (required >=13). Parsers must be compatible with runtime ABI.
```
## Parser/Features H L F I J
  - norg           x . x . x 

The following errors have been detected:
  - ERROR: norg(highlights): ...ejQS/usr/share/nvim/runtime/lua/vim/treesitter/query.lua:161: query: invalid node type at position 1
  - ERROR: norg(folds): ...ejQS/usr/share/nvim/runtime/lua/vim/treesitter/query.lua:161: query: invalid node type at position 4
  - ERROR: norg(injections): ...ejQS/usr/share/nvim/runtime/lua/vim/treesitter/query.lua:161: query: invalid node type at position 50
```

I'm running Ubuntu 20.04 LTS and I have just executed `:TSUpdate` and `:PackerSync`, but to no avail.

**Comment by vhyrro** [comment]
@Gelio the reason is because the new parser does not work with the main branch yet. I think what we should do is push the unstable branch to main. The issue is that I haven't written any real docs for the new stuff we're making yet and so I dunno whether that'll confuse people

**Comment by Gelio** [comment]
@vhyrro I see, thanks for explaining. In that case, maybe there should be an `unstable` branch of the parser as well for such changes, which would require using the `unstable` branch of the plugin? Just an idea, as otherwise, it looks like anyone updating the parser (or installing the plugin from scratch, which is what I've done) has to use the `unstable` branch

Just a thought for the future

**Comment by vhyrro** [comment]
@Gelio what we ended up doing is we merged the unstable branch into `main`. We also added an extra disclaimer in the README that Neorg is alpha software and that certain bits of documentation may be out of date. This should alleviate a bunch of issues that people may have had to this date 🤞

**Comment by AndOrangutan** [comment]
Just tried updating to the unstable branch and got this:
```
Error detected while processing BufEnter Autocommands for "*.norg":
E5108: Error executing lua ...t/neorg/lua/neorg/modules/core/norg/concealer/module.lua:362: attempt to call method 
'prev_named_sibling' (a nil value)
```
I am on Arch linux.

**Comment by glyh** [comment]
I have the same issue, hope this will be fixed soon

**Comment by JorisDeRidder** [comment]
I have still the same issue on OSX 10.15 and clang 12.0.0, even with the unstable branch. :TSInstall norg gives lots of errors. The first one is:
`src/scanner.cc:360:41: error: expected ';' at end of declaration list`
`    std::vector<size_t>& get_tag_stack() noexcept { return m_TagStack; }`

**Comment by danymat** [comment]
As this is the most upvoted answer: https://github.com/nvim-neorg/neorg/issues/74#issuecomment-906627223

I tried simplifying it a little bit.
We do not need to create symlinks for `cc` stuff, as tree sitter allow us to have an environment variable called `CC`.

You can do `CC=/path/to/gcc-12 nvim` before doing any actual update or installation, or put `CC=/path/to/gcc-12` in your `bashrc` or `zshrc` for the change to be permanent

**Comment by ray-x** [comment]
I followed instructions in https://github.com/nvim-neorg/neorg/issues/74#issuecomment-906627223
Still have issue.
double check with this:
```
gcc --version
gcc (Homebrew GCC 11.2.0_2) 11.2.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Also I linked `gcc`, linked `cc`, export `CC`, none of the tricks helps. Is there anything missing?

**Comment by vhyrro** [comment]
Hmm, i think that looks fine. @danymat since you're the one that fixed the issue is there anything else ray can do?

Man we really need that makefile support in the nvim-treesitter repo

**Comment by danymat** [comment]
> I followed instructions in https://github.com/nvim-neorg/neorg/issues/74#issuecomment-906627223
> 
> Still have issue.
> 
> double check with this:
> 
> ```
> 
> gcc --version
> 
> gcc (Homebrew GCC 11.2.0_2) 11.2.0
> 
> Copyright (C) 2021 Free Software Foundation, Inc.
> 
> This is free software; see the source for copying conditions.  There is NO
> 
> warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
> 
> ```
> 
> 
> 
> Also I linked `gcc`, linked `cc`, export `CC`, none of the tricks helps. Is there anything missing?

Hello, can you print the output of `which cc`, `env | GREP CC` in the same terminal just before you enter neovim ?

**Comment by ray-x** [comment]
Yes, here is what is printed out:

```
❯❯❯ cc -v
Using built-in specs.
COLLECT_GCC=/usr/local/Cellar/gcc/11.2.0_2/bin/gcc-11
COLLECT_LTO_WRAPPER=/usr/local/Cellar/gcc/11.2.0_2/libexec/gcc/x86_64-apple-darwin19/11/lto-wrapper
Target: x86_64-apple-darwin19
Configured with: ../configure --prefix=/usr/local/Cellar/gcc/11.2.0_2 --libdir=/usr/local/Cellar/gcc/11.2.0_2/lib/gcc/11 --disable-nls --enable-checking=release --with-gcc-major-version-only --enable-languages=c,c++,objc,obj-c++,fortran,d --program-suffix=-11 --with-gmp=/usr/local/opt/gmp --with-mpfr=/usr/local/opt/mpfr --with-mpc=/usr/local/opt/libmpc --with-isl=/usr/local/opt/isl --with-zstd=/usr/local/opt/zstd --with-pkgversion='Homebrew GCC 11.2.0_2' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --enable-libphobos --build=x86_64-apple-darwin19 --with-system-zlib --disable-multilib --with-native-system-header-dir=/usr/include --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.2.0 (Homebrew GCC 11.2.0_2) 


❯❯❯ echo $CC
/usr/local/Cellar/gcc/11.2.0_2

❯❯❯ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/local/Cellar/gcc/11.2.0_2/libexec/gcc/x86_64-apple-darwin19/11/lto-wrapper
Target: x86_64-apple-darwin19
Configured with: ../configure --prefix=/usr/local/Cellar/gcc/11.2.0_2 --libdir=/usr/local/Cellar/gcc/11.2.0_2/lib/gcc/11 --disable-nls --enable-checking=release --with-gcc-major-version-only --enable-languages=c,c++,objc,obj-c++,fortran,d --program-suffix=-11 --with-gmp=/usr/local/opt/gmp --with-mpfr=/usr/local/opt/mpfr --with-mpc=/usr/local/opt/libmpc --with-isl=/usr/local/opt/isl --with-zstd=/usr/local/opt/zstd --with-pkgversion='Homebrew GCC 11.2.0_2' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --enable-libphobos --build=x86_64-apple-darwin19 --with-system-zlib --disable-multilib --with-native-system-header-dir=/usr/include --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.2.0 (Homebrew GCC 11.2.0_2) 
```


later on, I found that it was alias setup not working. I was intend to use alias instead of symbol link for `cc` . Seems I have to symble link  `cc` to  `gcc-11` make things works.
If only setup `CC`, it does not work, also link `gcc` to `gcc-11` does not work.

**Comment by vitaly** [comment]
this worked for me: `CC=gcc-11 nvim` and then `:TSInstall norg`

**Comment by Cristian-L-Zambrano** [comment]
"export CC=gcc-11" on mac works for me.

**Comment by typkrft** [comment]
> Okay i solved the issue (I'm on Mac 11.4).
> 
> What causes the issue is that I didn't have the latest gcc installed. Below are the steps to solve:
> 
>     1. Install latest gcc: `brew install gcc`
> 
>     2. Check out which compiler is used in TSInstall: open nvim and run `:checkhealth nvim_treesitter`.
> 
> 
> Note: in installation header, I have this:
> 
> ```
> - OK: `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl" }
> ```
> 
> So that means tree sitter will compile with `cc`
> 
>     3. Alias brew gcc-11 with cc: do `brew info gcc`to see where is located the gcc binaries, for me it's:
> 
> 
> ```
> /usr/local/Cellar/gcc/11.2.0
> ```
> 
> I can now alias to some directory in my path like `/usr/local/bin`:
> 
> ```
> ln -s /usr/local/Cellar/gcc/11.2.0/bin/gcc-11 /usr/local/bin/cc
> ```
> 
> You can now restart Neovim and run `TSUpdate norg`

This does not seem to work for me.

1. `brew install gcc`
2. `ln -s /opt/homebrew/Cellar/gcc/11.2.0_3/bin/gcc-11 /Users/brandon/.config/bin/cc` 
3. `cc -v`
```bash
Using built-in specs.
COLLECT_GCC=cc
COLLECT_LTO_WRAPPER=/opt/homebrew/Cellar/gcc/11.2.0_3/bin/../libexec/gcc/aarch64-apple-darwin21/11/lto-wrapper
Target: aarch64-apple-darwin21
Configured with: ../configure --prefix=/opt/homebrew/opt/gcc --libdir=/opt/homebrew/opt/gcc/lib/gcc/11 --disable-nls --enable-checking=release --with-gcc-major-version-only --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-11 --with-gmp=/opt/homebrew/opt/gmp --with-mpfr=/opt/homebrew/opt/mpfr --with-mpc=/opt/homebrew/opt/libmpc --with-isl=/opt/homebrew/opt/isl --with-zstd=/opt/homebrew/opt/zstd --with-pkgversion='Homebrew GCC 11.2.0_3' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --build=aarch64-apple-darwin21 --with-system-zlib --with-native-system-header-dir=/usr/include --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.2.0 (Homebrew GCC 11.2.0_3) 
``` 
4. `nvim`
5. :TSInstall norg 
```
nvim-treesitter[norg]: Error during compilation
/bin/sh: -c: line 0: syntax error near unexpected token `)'
/bin/sh: -c: line 0: `cc parser.o scanner.o -o parser.so -shared -L/opt/homebrew/lib) -lstdc++'
make: *** [parser.so] Error 2
```

:checkhealth nvim_treesitter
```
nvim_treesitter: health#nvim_treesitter#check
========================================================================
## Installation
  - OK: `tree-sitter` found 0.20.2 (parser generator, only needed for :TSInstallFromGrammar)
  - OK: `node` found v17.4.0 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
    Version: cc (Homebrew GCC 11.2.0_3) 11.2.0
  - OK: Neovim was compiled with tree-sitter runtime ABI version 13 (required >=13). Parsers must be compatible with runtime ABI.
```

**Comment by brendonsoto** [comment]
For anyone running into this but receiving an error looking like,
```
ld: Assertion failed: (_file->_atomsArrayCount == computedAtomCount && "more atoms allocated than expected"), function parse, file macho_relocatable_file.cpp, line 2061.
collect2: error: ld returned 1 exit status
```
check out [this thread](https://github.com/iains/gcc-12-branch/issues/6#issuecomment-1266660607).
tl;dr - either downgrade command line tools for xcode or dl the beta.
I hope this saves someone some time.

**Comment by haikyuu** [comment]
Thanks @brendonsoto facing the same issue

**Comment by transcendr** [comment]
@danymat  thanks worked!

I did:
`which gcc-12` which gave me `/opt/homebrew/bin/gcc-12`

Then in `.zshrc`:

`export CC=/opt/homebrew/bin/gcc-12`

Then I ran:

`nvim -c "TSInstallSync norg"`

And all good!

**Comment by howarddo2208** [comment]
I just got this error during a fresh install on MacOS. I think the README instruction is misleading, what needed is gcc11, not C++14 at all.
